routemaster/models/batch.rb

```diff
@@ -1,56 +1,245 @@
+      end
+
+
+      # Transitions a batch from current.
+      # A non-current batch will no have data added to it.
+      # 
+      # It is not an error if the batch is not current, or no longer exists.
+      def promote
+        _redis_lua_run(
+          'batch_promote',
+          keys: [_batch_ref_key],
+          argv: [@uid])
         self
       end
 
-      def events
-        @batch.select { |m| m.kind_of?(Event) }
+
+      # Removes all references to the batch (it's been delivered, or autodropped)
+      def delete
+        count = _redis_lua_run(
+          'batch_delete',
+          keys: [_batch_key, _index_key, _batch_ref_key, _batch_counter_key, _event_counter_key],
+          argv: [@uid, PREFIX_COUNT, _subscriber_name])
+        broadcast(:events_removed, name: _subscriber_name, count: count)
+        self
       end
 
-      def age
-        now = Routemaster.now
-        @batch.map { |m| now - m.timestamp }.max || 0
+
+      module ClassMethods
+        include Wisper::Publisher
+
+        # Add the data to the subscriber's current batch. A new batch will be
+        # created as needed. The batch will be promoted if it's full.
+        def ingest(data:, timestamp:, subscriber:)
+          batch_ref_key = _batch_ref_key(subscriber.name)
+          now           = Routemaster.now
+          deadline      = timestamp + subscriber.timeout
+
+          # Ingestion might create a new batch if there is no current batch (pointed to by
+          # `batch_ref_key`) changes between the SRANDMEMBER and the EVAL.
+          # To this effect, we provide an alternate batch UID to be used if
+          # creating a batch.
+          uid = _redis.srandmember(batch_ref_key)
+          alt_uid = _generate_uid
+
+          yield if block_given? # this is used in tests only, to inject behaviour to simulate concurrency
+
+          actual_uid =  _redis_lua_run(
+              'batch_ingest',
+              keys: [batch_ref_key, _batch_key(uid), _batch_key(alt_uid), _index_key, _batch_counter_key, _event_counter_key],
+              argv: [uid, alt_uid, data, subscriber.name, PREFIX_COUNT, subscriber.max_events, now])
+          
+          broadcast(:events_added, name: subscriber.name, count: 1)
+          new(subscriber: subscriber, uid: actual_uid, deadline: deadline)
+        end
+
+
+        def counters
+          {
+            batches: _redis.hgetall(_batch_counter_key).map_values(&:to_i).tap { |h| h.default = 0 },
+            events:  _redis.hgetall(_event_counter_key).map_values(&:to_i).tap { |h| h.default = 0 },
+          }
+        end
+
+
+        def scrub
+          # TODO:
+          # iterate over all batches so the caller can enqueue them
+        end
+
+
+        def all
+          Iterator.new
+        end
+
+
+        private
+
+        def _generate_uid
+          SecureRandom.urlsafe_base64(15)
+        end
+
+
+        def _batch_ref_key(name)
+          name ? "batches:current:#{name}" : nil
+        end
+
```

Comment:
I was more worried about the fact that with unshareded we are binding ourselves to 1 redis instance meaning everything is slow as soon as one operation is slow. In general do we keep sharding in mind here? spreading multiple queues over multiple redissss (what is the plural of redis 😁 ) seems reasonable. (there are more keys which seem global thats why I'm concerned)